### DISCUSSION TITLE
"Found no way to assign variables"

### DISCUSSION THREAD

**Discussion by rvalek** [EvgSkv/logica#208] [Q&A] [answered]
Hey Evgeniy!

I'm trying to understand the aggregation behaviour and the errors I'm getting. The following snippet works fine:
```
Test(matching: MatchingMember(["ab", "ac", "bc"], ["a%", "%"]));

MatchingMember(items, patterns) = member :-
  member in items,
  p in patterns,
  Like(member, p);
```

However, if I make `MatchingMember` `disctinct`, or change it to functional aggregation (`List= member :-`), it no longer compiles:
`[ Error ] Found no way to assign variables: items, patterns. This error might also come from injected sub-rules.`

Where is this coming from? Is there something I can do better?

Thank you!
~Robert

**Comment by rvalek** [comment]
I found a workaround to get what I want:
```

Test(matching: AllMatching(["ab", "ac", "bc"], ["a%", "%"]));

MatchingMember(items, patterns, member) = member :-
  member in items,
  p in patterns,
  Like(member, p);

AllMatching(items, patterns) = members :-
  members Set= MatchingMember(items, patterns);
```

But if I change `AllMatching` predicate to be more succinct, it no longer compiles:
`AllMatching(items, patterns) Set= MatchingMember(items, patterns);`

**Comment by EvgSkv** [comment, marked as answer]
Hi @rvalek ,
yes, this works as intended and your solution is optimal.
We do need to document this better. For now I have added a [brief section](https://colab.research.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb#scrollTo=q55qN2URLwu-) into the tutorial.

Logica tries to have a simple compiler, so user can easily control the SQL query and thus optimize their program.

So any predicate must be _injectible_ or _concrete_.
Predicate is injectible if it is defined with a *single non-aggregating rule*. Such predicate is allowed to use aggregating expressions. This is exactly how you defined `AllMatching` predicate in your fix.
Injectable predicates can be "injected" into predicates that call them via substitution.

Concrete predicate can be compiled into SQL itself.

The way `AllMatching` is defined in your original post is neither injectable, nor concrete, so Logica can't compile it.
Since it uses aggregation it assumes that it's concrete and tries to compile into SQL stand alone, hence the error.

It is quite possible that in the future we may want to make compiler more flexible and compile programs like your original, but it wasn't done yet.

Hopefully this answer makes sense. Let me know if you have further questions.

**Reply by rvalek** [reply]
@EvgSkv, I really appreciate the detailed explanation. It makes sense.
Thank you for your great work on Logica, I keep being impressed as I use and discover it more. I think you're cool!

Robert

**Comment by EvgSkv** [comment]
@rvalek thank you very much for your supportive feedback! Don't hesitate to reach out if you have further questions.
